### ISSUE TITLE
YogaKit does not correctly account for composite UIView when doing layout

### ISSUE THREAD

**Issue by guidomb** [facebook/yoga#290]
I've been trying to render an instance of `MKMapView` but failed to do so.

Here is a code that works **without** using Yoga

```swift
import UIKit
import MapKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        let m = MKMapView()
        m.frame.size = CGSize(width: self.view.frame.size.width, height: 100)
        self.view.addSubview(m)
    }

}
```

which renders the following view

![simulator screen shot dec 19 2016 1 24 12 am](https://cloud.githubusercontent.com/assets/390645/21301187/fdfbcc02-c589-11e6-8549-47c9c74d9584.png)

but when I try to render a same map view using Yoga with the following code

```swift
import UIKit
import MapKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        let m = MKMapView()

        self.view.yg_usesYoga = true
        self.view.yg_setWidth(self.view.bounds.size.width)
        self.view.yg_setHeight(self.view.bounds.size.height)
        
        m.yg_usesYoga = true
        m.yg_setWidth(self.view.bounds.size.width)
        m.yg_setHeight(100)
        
        self.view.addSubview(m)
        self.view.yg_applyLayout()
    }

}
```

I get this 

![simulator screen shot dec 19 2016 1 26 10 am](https://cloud.githubusercontent.com/assets/390645/21301203/26ca417c-c58a-11e6-912c-c9707669fb9e.png)

Which is basically a "white" map

**Comment by emilsjolander** [comment]
Is the "legal" text part of MKMapView?

**Comment by guidomb** [comment]
Yes it is.

**Comment by emilsjolander** [comment]
Yes ok. This is an issue I found last week. MKMapView is a composite view which is not correctly handled in the YogKit implementation. MKMapView should be treated as a leaf node in the Yoga node tree but due to it being marked with `usesYoga` and it having children it does not get treated that way and this bug happens. For no you can fix this by not marking the map as using yoga and just setting a frame on it directly. UIButton and similar views have the same issue.

**Comment by guidomb** [comment]
OK thanks for the quick response. I'll try that. I haven't experienced this issue with UIButton though. Is there anything I can do to help fixing this bug? Not very familiar for Yoga's implementation but maybe you can point me to piece of code that might be causing this issue.

**Comment by emilsjolander** [comment]
I discussed this with a colleague the other day. We didn't come up with a solution we were super happy with which is why I didn't make any change. The only solution we came up with is to have a property called `yg_isLeaf` which can be set on certain views. This is the line of code which would check for `yg_isLeaf` https://github.com/facebook/yoga/blob/master/YogaKit/UIView%2BYoga.m#L283

**Comment by guidomb** [comment]
@emilsjolander If you want I can add private method like 

```objc
- (BOOL)yg_isLeaf {
  NSNumber * isLeaf = objc_getAssociatedObject(self, &kYGIsLeafAssociatedObjectKey);
  return [isLeaf boolValue] || ![view yg_usesYoga] || view.subviews.count == 0;
}
```

and then make a category for `MKMapView` that sets the associated object to `@YES`.

Not the cleanest solution but it should work. What do you think?

**Comment by emilsjolander** [comment]
This is not scaleable. I think a better solution is to just add a prop to
the existing category and then you can create your own subclasses or
categories which set that prop.

**Comment by guidomb** [comment]
You mean use the same mechanism as how, for example, `yg_setWidth` is implemented? Add a `yg_setIsLeaf` and then make a category in `MKMapView` which always return `YES`?. 

If that is the case I wonder if category extension have a well defined "method order". By this I mean if a category method defined on a subclass has precedence over the same category method defined on a super class. I would bet that this is undefined behavior.

**Comment by emilsjolander** [comment]
@guidomb I think you could add a category method which is `-(instancetype)initForYoga` which calls that setter. Or your category could swizzle the implementation of `-(BOOL)isLeaf` on `MKMapView`

**Comment by guidomb** [comment]
Before I find the best way to do such thing I need to understand this change a little better because I've made a quick hack to check if this would actually fix the issue and I am getting a crash. I've changed the line you pointed before to

```objc
static void YGAttachNodesFromViewHierachy(UIView *view) {
  YGNodeRef node = [view ygNode];

  // Only leaf nodes should have a measure function
  if ([view isKindOfClass:[MKMapView class]] || ![view yg_usesYoga] || view.subviews.count == 0) {
    YGNodeSetMeasureFunc(node, YGMeasureView);
    YGRemoveAllChildren(node);
  } else {
    YGNodeSetMeasureFunc(node, NULL);


 // ... the remaining of the function's code
}
```

which successfully forces `YGAttachNodesFromViewHierachy` to call `YGNodeSetMeasureFunc` for the map view by adding `[view isKindOfClass:[MKMapView class]]` to if expression. The problem is that this generates a crash in https://github.com/facebook/yoga/blob/master/YogaKit/UIView%2BYoga.m#L360

The crash being

```
2016-12-19 10:38:43.728 Portal[25806:11779390] *** Terminating app due to uncaught exception 'CALayerInvalidGeometry', reason: 'CALayer position contains NaN: [nan nan]'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010634034b __exceptionPreprocess + 171
	1   libobjc.A.dylib                     0x0000000105da121e objc_exception_throw + 48
	2   CoreFoundation                      0x00000001063a9265 +[NSException raise:format:] + 197
	3   QuartzCore                          0x000000010d1b12da _ZN2CA5Layer12set_positionERKNS_4Vec2IdEEb + 152
	4   QuartzCore                          0x000000010d1b144f -[CALayer setPosition:] + 44
	5   QuartzCore                          0x000000010d1b1aba -[CALayer setFrame:] + 650
	6   UIKit                               0x0000000106baf694 -[UIView(Geometry) setFrame:] + 854
	7   Portal                              0x00000001056cf981 YGApplyLayoutToViewHierarchy + 945
	8   Portal                              0x00000001056cfb06 YGApplyLayoutToViewHierarchy + 1334
	9   Portal                              0x00000001056cfb06 YGApplyLayoutToViewHierarchy + 1334
	10  Portal                              0x00000001056cfb06 YGApplyLayoutToViewHierarchy + 1334
	11  Portal                              0x00000001056cfb06 YGApplyLayoutToViewHierarchy + 1334
	12  Portal                              0x00000001056cf5ca -[UIView(Yoga) yg_applyLayout] + 122
	13  Portal                              0x00000001056de673 _TFV6Portal16YogaLayoutEngine6layoutfT4viewCSo6UIView4withVS_6Layout6insideS1__T_ + 4019
	14  Portal                              0x00000001056e3599 _TTWV6Portal16YogaLayoutEngineS_12LayoutEngineS_FS1_6layoutfT4viewCSo6UIView4withVS_6Layout6insideS2__T_ + 2473
	15  Portal                              0x00000001056f2c95 _TFV6Portal13UIKitRenderer6renderu1_R0_S_9Componentxzw0_14PropertiesType_zw0_16CustomStyleSheetrfT8viewNodeq0__T_ + 3653
	16  Portal                              0x00000001056e941b _TFC6Portal14ViewController11viewDidLoadfT_T_ + 13067
	17  Portal                              0x00000001056e99e2 _TToFC6Portal14ViewController11viewDidLoadfT_T_ + 34
	18  UIKit                               0x0000000106cabc99 -[UIViewController loadViewIfRequired] + 1258
	19  UIKit                               0x0000000106cac0cc -[UIViewController view] + 27
	20  UIKit                               0x0000000106b75c51 -[UIWindow addRootViewControllerViewIfPossible] + 71
	21  UIKit                               0x0000000106b763a2 -[UIWindow _setHidden:forced:] + 293
	22  UIKit                               0x0000000106b89cb5 -[UIWindow makeKeyAndVisible] + 42
	23  UIKit                               0x0000000106b02c89 -[UIApplication _callInitializationDelegatesForMainScene:transitionContext:] + 4818
	24  UIKit                               0x0000000106b08de9 -[UIApplication _runWithMainScene:transitionContext:completion:] + 1731
	25  UIKit                               0x0000000106b05f69 -[UIApplication workspaceDidEndTransaction:] + 188
	26  FrontBoardServices                  0x000000010f41b723 __FBSSERIALQUEUE_IS_CALLING_OUT_TO_A_BLOCK__ + 24
	27  FrontBoardServices                  0x000000010f41b59c -[FBSSerialQueue _performNext] + 189
	28  FrontBoardServices                  0x000000010f41b925 -[FBSSerialQueue _performNextFromRunLoopSource] + 45
	29  CoreFoundation                      0x00000001062e5311 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	30  CoreFoundation                      0x00000001062ca59c __CFRunLoopDoSources0 + 556
	31  CoreFoundation                      0x00000001062c9a86 __CFRunLoopRun + 918
	32  CoreFoundation                      0x00000001062c9494 CFRunLoopRunSpecific + 420
	33  UIKit                               0x0000000106b047e6 -[UIApplication _run] + 434
	34  UIKit                               0x0000000106b0a964 UIApplicationMain + 159
	35  Portal                              0x00000001056eaa7f main + 111
	36  libdyld.dylib                       0x0000000109d7c68d start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
(lldb) 
```

**Comment by emilsjolander** [comment]
You also need to update https://github.com/facebook/yoga/blob/master/YogaKit/UIView%2BYoga.m#L371

**Comment by guidomb** [comment]
@emilsjolander Oops! sorry didn't see that. You were right it worked! Now the maps renders correctly. I'll prepare a PR. Thanks a lot helping me!

**Comment by guidomb** [comment]
I've created https://github.com/facebook/yoga/pull/291 we can continue the discussion there.

**Comment by dshahidehpour** [comment]
@guidomb Thanks for using YogaKit! I'm Dustin, and I'm one of the main creators of YogaKit. I ran into this bug a few days and have been working on a fix internally at FB. 

The TL:DR; is that I'm landing a fix this morning. It will sync with this repo shortly after.

As you correctly spotted, add YogaKit properties to a composite views removes it use of the `MeasureFunc` which usually results in crashes, and a destroyed layout.

Before we would determine a leaf like so:
```
const BOOL isLeaf = ![view yg_usesYoga] || view.subviews.count == 0;
```

This would be incorrect for composite views (like `MKMapView`) because it contains subviews, but, we only want the yoga properties to affect the view as a whole.

So here's my solution:
```
- (BOOL)yg_isLeaf {
  NSAssert([NSThread isMainThread], @"We are accessing subviews....");
  if ([self yg_usesYoga]) {
      for (UIView *subview in self.subviews) {
          if ([subview yg_usesYoga] && [subview yg_includeInLayout]) {
              return NO;
          }
      }
  }

  return YES;
}
```
Now, when we check to see if something is a leaf, it can have subviews, as long as none of them use yoga! :)

**Comment by guidomb** [comment]
@dshahidehpour Cool! Thanks for sharing this! Looking forward for this fix!

**Comment by dshahidehpour** [comment]
The fix for this landed in 835bb1c. I'm going to close this, but feel free to open if something else arises!
